### PR TITLE
Implement RemoteEmuMPSEmulator

### PR DESCRIPTION
**Pull request by Yoric**

### CODE REVIEW

File: qek/data/extractors.py [hunk 1]
```diff
@@ -645,8 +663,14 @@ def _ingest(self, batches: dict[str, Batch]) -> None:
             assert len(batch.jobs) == 1
             for _, job in batch.jobs.items():
```
Comment (RolandMacDoland) [inline qek/data/extractors.py:664]:
```suggestion
            for job in batch.jobs.values():
```

File: qek/data/extractors.py [hunk 2]
```diff
@@ -645,8 +663,14 @@ def _ingest(self, batches: dict[str, Batch]) -> None:
...
                 if job.status == "DONE":
-                    state_dict = job.result
-                    assert state_dict is not None
+                    state_dict = self._state_extractor(job, compiled.sequence)
+                    if state_dict is None:
+                        logger.warning(
+                            "Batch %s (graph %s) did not return a usable state, skipping",
```
Comment (RolandMacDoland) [inline qek/data/extractors.py:669]: What's the rationale again for not using string formatting here ?

File: qek/data/extractors.py [hunk 3]
```diff
@@ -698,14 +722,18 @@ def states(self) -> list[dict[str, int]]:
         return self._results.states
 
 
-class QPUExtractor(BaseExtractor[GraphType]):
+class BaseRemoteExtractor(BaseExtractor[GraphType], Generic[GraphType]):
     """
-    A Extractor that uses a distant physical QPU to run sequences
-    compiled from graphs.
+    An Extractor that uses a distant a Quantum Device published
```
Comment (RolandMacDoland) [inline qek/data/extractors.py:727]:
```suggestion
    An Extractor that uses a remote Quantum Device published
```

File: qek/data/extractors.py [hunk 4]
```diff
@@ -698,14 +722,18 @@ def states(self) -> list[dict[str, int]]:
...
+    through Pasqal Cloud, to run sequences compiled from graphs.
```
Comment (RolandMacDoland) [inline qek/data/extractors.py:728]:
```suggestion
    on Pasqal Cloud, to run compiled sequences from graphs.
```

File: qek/data/extractors.py [hunk 5]
```diff
@@ -745,13 +773,33 @@ def __init__(
...
+        raise Exception("Not implemented")
+
+    def _run(
+        self,
+        state_extractor: Callable[[Job, pl.Sequence], dict[str, int] | None],
+        emulator: EmulatorType | None,
+        config: BaseConfig | None,
+    ) -> PasqalCloudExtracted:
         if len(self.sequences) == 0:
             logger.warning("No sequences to run, did you forget to call compile()?")
```
Comment (RolandMacDoland) [inline qek/data/extractors.py:792]: Why just logging a warning and letting it through ?

File: qek/data/extractors.py [hunk 6]
```diff
@@ -778,6 +828,134 @@ def run(self) -> CloudExtracted:
...
+        batch_id: Use this to resume a workflow e.g. after turning off
+            your computer while the QPU was executing your sequences.
+    """
+
+    def __init__(
+        self,
+        compiler: BaseGraphCompiler[GraphType],
+        project_id: str,
+        username: str,
+        device_name: str = "FRESNEL",
```
Comment (RolandMacDoland) [inline qek/data/extractors.py:928]: It doesn't seem the right ordering (wrt the extractor above). And should the `device_name` still be defaulted here ?

**Review by RolandMacDoland — COMMENTED** [review]
Thanks @Yoric, some comments from my side.

File: qek/data/extractors.py [hunk 7]
```diff
@@ -548,13 +550,29 @@ class CloudExtracted(BaseExtracted):
     """
 
     def __init__(
-        self, compiled: list[Compiled], batch_ids: list[str], sdk: SDK, path: str | None = None
+        self,
+        compiled: list[Compiled],
+        batch_ids: list[str],
+        sdk: SDK,
+        state_extractor: Callable[[Job, pl.Sequence], dict[str, int] | None],
+        path: str | None = None,
```
[the remark this answers is not in the document]
Comment (Yoric) [reply qek/data/extractors.py:558]: Should I confess that I didn't know people were using it?

File: qek/data/extractors.py (hunk 2 above)
Comment (Yoric) [reply qek/data/extractors.py:669]: Generally a good practice in logging libraries: don't build strings before knowing whether they'll actually be displayed.

File: qek/data/extractors.py (hunk 5 above)
Comment (Yoric) [reply qek/data/extractors.py:792]: What would you suggest? Raising an exception?

Comment (RolandMacDoland) [reply qek/data/extractors.py:792]: Idk, do we continue if there isn't anything to execute ? In some sense, we can't prevent users from running empty jobs.

File: qek/data/extractors.py [hunk 8]
```diff
@@ -778,6 +828,134 @@ def run(self) -> CloudExtracted:
...
+        )
+
+    def run(self) -> PasqalCloudExtracted:
+        return self._run(emulator=None, config=None, state_extractor=lambda job, _seq: job.result)
+
+
+class RemoteEmuMPSExtractor(BaseRemoteExtractor[GraphType]):
+    """
+    An Extractor that uses the high-performance distributed emu-mps emulator
+    through Pasqal Cloud, to run sequences compiled from graphs.
```
[the remark this answers is not in the document]
Comment (Yoric) [reply qek/data/extractors.py:899]: Yes, that's the idea. We have _very_ beefy GPUs on the cloud, not everybody has the same at home :)

File: qek/data/extractors.py (hunk 6 above)
Comment (Yoric) [reply qek/data/extractors.py:929]: Ordering fixed.

I believe that `device_name` should be defaulted, too. I expect that our users will not know the list of devices from the top of their head.

File: qek/data/extractors.py [hunk 9]
```diff
@@ -745,13 +774,33 @@ def __init__(
...
 
-    def run(self) -> CloudExtracted:
+    @abc.abstractmethod
+    def run(
+        self,
+    ) -> PasqalCloudExtracted:
+        """
+        Launch the extraction.
+        """
+        raise Exception("Not implemented")
```
Comment (RolandMacDoland) [inline qek/data/extractors.py:784]:
```suggestion
        raise NotImplementedError()
```

File: qek/data/extractors.py [hunk 10]
```diff
@@ -778,6 +829,134 @@ def run(self) -> CloudExtracted:
...
+            project_id=project_id,
+            username=username,
+            device_name=device_name,
+            password=password,
+            batch_ids=batch_ids,
+            path=path,
+        )
+
+    def run(self, dt: int = 10) -> PasqalCloudExtracted:
+        def extractor(job: Job, sequence: pl.Sequence) -> dict[str, int] | None:
```
Comment (RolandMacDoland) [inline qek/data/extractors.py:945]: Does this callback has to return `None` ? Wouldn't an empty `dict` suffice ?

**Review by RolandMacDoland — APPROVED** [review]
Thanks @Yoric. Minor comments otherwise LGTM.

File: qek/data/extractors.py (hunk 10 above)
Comment (Yoric) [reply qek/data/extractors.py:945]: Mmmmh... The `None` case means that the job somehow didn't return a result, while the empty `dict` means that the job returned an empty result. In practice, neither should happen, but just in case, I don't see any major reason to mask the difference.

Comment (RolandMacDoland) [reply qek/data/extractors.py:945]: Well, if it didn't return a result, it means there has been an issue somewhere. In that case, an exception should have been raised. IMHO, `None` creates more confusion because it is neither a result nor an exception.
